extras/sht3x/sht3x.c

```diff
@@ -0,0 +1,385 @@
+ 
+    uint32_t            bg_period;          // the measurement period
+    TaskHandle_t        bg_task;            // the measurement background task
+
+    sht3x_value_set_t   values;             // last measured sensor values
+    
+    sht3x_cb_function_t user_cb_function;   // user callback function for passing results
+
+    TaskHandle_t        user_task;          // user task waiting to be notified on new results
+    uint32_t            user_notify_value;  // notification value evaluated by the user task
+    
+} sht3x_sensor_t;
+
+
+sht3x_sensor_t sht3x_sensors[SHT3x_MAX_SENSORS];
+
+
+static bool sht3x_send_command(uint32_t id, uint16_t cmd)
+{
+    uint8_t data[2] = { cmd / 256, cmd % 256 };
+
+    debug("send MSB command byte %02x", __FUNCTION__, data[0]);
+    debug("send LSB command byte %02x", __FUNCTION__, data[1]);
+
+    int err = i2c_slave_write(sht3x_sensors[id].bus, sht3x_sensors[id].addr, 0, data, 2);
+  
+    if (err)
+    {
+        error_addr("error %d on write command %02x", 
+                   __FUNCTION__, sht3x_sensors[id].bus, sht3x_sensors[id].addr, err, cmd);
+        return false;
+    }
+      
+    return true;
+}
+
+
+static bool sht3x_read_data(uint32_t id, uint8_t *data,  uint32_t len)
+{
+    int err = i2c_slave_read(sht3x_sensors[id].bus, sht3x_sensors[id].addr, 0, data, len);
+
+    if (err)
+    {
+        error_addr("error %d on read %d byte", 
+                   __FUNCTION__, sht3x_sensors[id].bus, sht3x_sensors[id].addr, err, len);
+        return false;
+    }
+
+#   ifdef SHT3x_DEBUG
+    printf("SHT3x: %s: Read following bytes: ", __FUNCTION__);
+    for (int i=0; i < len; i++)
+        printf("%02x ", data[i]);
+    printf("\n");
+#   endif
+
+    return true;
+}
+
+
+static bool sht3x_is_available (uint32_t sensor)
+{
+    uint8_t data[3];
+
+    if (!sht3x_send_command(sensor, SHT3x_STATUS_CMD))
+    {
+	    debug("called from", __FUNCTION__);
+        return false;
+    }
+        
+    if (!sht3x_read_data(sensor, data, 3))
+    {
+	    debug("Called from", __FUNCTION__);
+        return false;
+    }
+
+    return true;
+}
+
+
+static bool sht3x_valid_sensor (uint32_t sensor, const char* function)
+{
+    if (sensor < 0 || sensor > SHT3x_MAX_SENSORS)
+    {
+        debug("wrong sensor id %d.", function, sensor);
+        return false;
+    }
+    
+    if (!sht3x_sensors[sensor].active)
+    {
+        debug("sensor with id %d is not active.", function, sensor);
+        return false;
+    }
+    
+    return true;
+}
+
+
+static void sht3x_compute_values (uint8_t id, uint8_t* data)
+{
+  sht3x_value_set_t  values;
+
+  values.temperature_c = ((((data[0] * 256.0) + data[1]) * 175) / 65535.0) - 45;
+  values.temperature_f = ((((data[0] * 256.0) + data[1]) * 347) / 65535.0) - 49;
+  values.humidity      = ((((data[3] * 256.0) + data[4]) * 100) / 65535.0);
+  
+  sht3x_sensors[id].values = values;
+}
+
+
+static void sht3x_background_task (void *pvParameters)
```

Comment:
Personally I think this is the wrong level of abstraction for a device driver, and would move the task into the example code. There are often more than one device on an i2c bus, and the tasks created here can not coordinate these so it so limited that it seems out of place here. Any substantive use will need a separate task anyway, so suggest just keeping the code needed by such tasks here and moving the rest into the examples. Then you could eliminate the callback and task notification and multiple sensor abstractions from the code here.

Reply:
@ourairquality 
Many thanks for your review. Your suggestion would make the life much more easier for me since I would have to document much less. But, moving the code from the background task into a user task would mean that the user is completly responsible to implement the protocol and the timings of the sensor. From my point of view, the user task should not take care about this. I would expect from a driver that I just call a function to get the values without thinking about details like protocol timings and so on. 

The driver can be used exactly in that way. Let the background task doing measurements periodically and call simply function sht3x_get_values whenever you need the values. You have not to take care about anything else.

I don't understand exactly your concern with multiple I2C device at the same bus. As I understand, the i2c driver takes care about concurrency situations.

Reply:
The i2c driver does not coordinate multiple tasks, so the caller is responsible, and as noted if this sht3x driver were to be used with any other devices on the bus this driver would not work anyway.

On the bmp180, I don't think that was a good abstraction either, it has the same problem, but it is layered. I don't use the higher level abstractions there either, but it has the `bmp180_measure` interface that does all the work and can be used.

For the timing, why not just block while measuring. The thread task does this anyway, so the polling measurement could be moved to a separate function, and the rest moved out.

Reply:
Ok, at least the i2c driver returns an busy error if another i2c device is communicating. I could improve the access to the i2c bus in that way, that the driver handles the busy situation and waits for a certain time. I think, this could be solved.

Yes, bmp180 driver has a single function for measurements. But, it starts the measurement and then uses sdk_os_delay_us to wait up to 30 ms before it reads the sensor values. I don't think that it is a good idea to block the system for 30 ms. Event though vTaskDelay could be used in such a function, the driver has no idea what task is delayed. For example, if the call of this single measurement function comes from a software timer callback, the timer service task would be delayed for 30 ms, ..., a bad idea.

Exactly that was the motivation to think about the solution with background task that is doing the measurement without disturbing other tasks or the system itself.

I wrote a driver for BME680 in same way, you saw it already in pull request #450. This sensor has a much more complex timings. The measurement cycle can vary from some milliseconds up to 4.5 seconds.

Ok, one possibility would be to separate the low level and the high level functions into seperate layers, but leaving the user alone with complex measurement procedures, ... I'm not sure whether its a good idea.

Reply:
I does not seem reliable enough to me to just depend on the i2c driver falling over if there are thread conflicts. It appears to be an attempt to use a protocol designed to handle conflict on the wire to manage thread conflicts at the driver level. I suspect there could be thread conflicts internally, such as access to the I/O registers etc that also need to be managed. I am not sure, but would not use code that did not coordinate thread access in the i2c driver. This might be done per-bus.

The delay in the bmp180 driver does not block the system, it can be pre-empted, and as you note vTaskDelay would have been a better choice there.

Your own code does not use a callback, it's all inline in the thread body, and that would be the usage. If people want to build higher level abstractions then I would leave that up to them in their code, and there are so many options for that.

I have not looked your bmp680 code yet, but the existing bmp280 code has the function `bmp280_read_fixed` that does all the work there too, no need for all the higher level baggage.

These are just my observations. Personally your driver would be unusable for me for the reasons noted, but perhaps it suits someone else.

Reply:
@ourairquality 
What about an low level interface like this one? 
```
void user_task(void *pvParameters)
{
    sht3x_values_t* results;
    int32_t duration;

    while (1) 
    {
        duration = sht3x_start_measurement (sensor);

        vTaskDelay (duration/portTICK_PERIOD_MS);

        if ((results = sht3x_get_results (sensor)))
            ... use the values

        vTaskDelay ((1000 - duration) / portTICK_PERIOD_MS);
    }
}
```
Function **sht3x_start_measurement** triggers the sensor to start the measurement and returns an estimated duration. **sht3x_get_results** reads raw sensor data and returns computed sensor values. I think, this would be easy and controllable enough.